src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Windows.cs

```diff
@@ -1,18 +1,134 @@
+            }
+
+            lock (SyncObj)
+            {
+                switch (signal)
+                {
+                    case PosixSignal.SIGINT:
+                    case PosixSignal.SIGQUIT:
+                    case PosixSignal.SIGTERM:
+                    case PosixSignal.SIGHUP:
+                        break;
+
+                    default:
+                        throw new PlatformNotSupportedException();
+                }
+
+                if (s_handlers.Count == 0 &&
+                    !Interop.Kernel32.SetConsoleCtrlHandler(s_handlerRoutine, addOrRemove: true))
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error();
+                }
+
+                var token = new Token(signal, handler);
+                s_handlers.Add(token);
+                return new PosixSignalRegistration(token);
+            }
+        }
+
+        public partial void Dispose()
+        {
+            lock (SyncObj)
+            {
+                if (_token is Token token)
+                {
+                    _token = null;
+
+                    s_handlers.Remove(token);
+                    if (s_handlers.Count == 0 &&
+                        !Interop.Kernel32.SetConsoleCtrlHandler(s_handlerRoutine, addOrRemove: false))
+                    {
+                        throw Win32Marshal.GetExceptionForLastWin32Error();
+                    }
+                }
+            }
+        }
+
+        private static bool HandlerRoutine(int dwCtrlType)
+        {
+            PosixSignal signal;
+            switch (dwCtrlType)
+            {
+                case Interop.Kernel32.CTRL_C_EVENT:
+                    signal = PosixSignal.SIGINT;
+                    break;
+
+                case Interop.Kernel32.CTRL_BREAK_EVENT:
+                    signal = PosixSignal.SIGQUIT;
+                    break;
+
+                case Interop.Kernel32.CTRL_SHUTDOWN_EVENT:
+                    signal = PosixSignal.SIGTERM;
+                    break;
+
+                case Interop.Kernel32.CTRL_CLOSE_EVENT:
+                    signal = PosixSignal.SIGHUP;
+                    break;
+
+                default:
+                    return false;
+            }
+
+            List<Token>? tokens = null;
+            lock (SyncObj)
+            {
+                foreach (Token token in s_handlers)
+                {
+                    if (token.Signal == signal)
+                    {
+                        (tokens ??= new()).Add(token);
+                    }
+                }
+            }
+
+            if (tokens is null)
+            {
+                return false;
+            }
+
+            var context = new PosixSignalContext(signal);
+            foreach (Token handler in tokens)
+            {
+                handler.Handler(context);
+            }
+
+            return context.Cancel;
+        }
+
+        private sealed class Token
```

Comment:
Is this an alternative to WeakReferences used by the Unix implementation? It would be nice if both Windows and Unix used similar implementation strategy. I like this implementation strategy better than the WeakReferences.

Reply:
> Is this an alternative to WeakReferences used by the Unix implementation?

Yes

> It would be nice if both Windows and Unix used similar implementation strategy. I like this implementation strategy better than the WeakReferences.

I agree. I'd started refactoring the Unix implementation, but then reverted it and decided to leave it for a subsequent PR, once all the functionality was in place across OSes.

Reply:
I like it better too.

Maybe add a comment that `Token` is to avoid rooting the `PosixSignalContext`.